main/src/main/scala/sbt/ScopeFilter.scala

```diff
@@ -9,29 +9,54 @@
 package sbt
 
 import sbt.internal.{ Load, LoadedBuildUnit }
-import sbt.internal.util.{ AttributeKey, Dag, Types }
+import sbt.internal.util.{ AttributeKey, Dag }
 import sbt.librarymanagement.{ ConfigRef, Configuration }
-import Types.const
+import sbt.internal.util.Types.const
 import Def.Initialize
 import sbt.Project.inScope
 import java.net.URI
 
+sealed abstract class ScopeFilter { self =>
+
+  /** Implements this filter. */
+  private[ScopeFilter] def apply(data: ScopeFilter.Data): Set[Scope]
```

Comment:
The old signature was: `def apply(data: ScopeFilter.Data): Scope => Boolean`. With the new signature we don't need to iterate on all scopes one by one. We can instead select the matching scopes directly from the `Map[Reference, Map[Config, Map[AtrributeKey, Set[Scope]]]]`